doc/languages-frameworks/agda.section.md

```diff
@@ -0,0 +1,80 @@
+---
+title: Agda
+author: Alex Rice (alexarice)
+date: 2020-01-06
+---
+# Agda
+
+## How to use Agda
+
+Agda can be installed with `agda.Agda`. For example:
+```
+$ nix-env -iA agda.Agda
+```
+
+To use agda with libraries, the `withPackages` function can be used. This function either takes:
++ A list of packages,
++ or a function which returns a list of packages when given the `agda` attribute set.
+
+For example, suppose we wanted a version of agda which has access to the standard library. This can be obtained with the expressions:
+
+```
+agda.withPackages [ agda.standard-library ]
+```
+
+or
+
+```
+agda.withPackages (p: [ p.standard-library ])
+```
+
+If you want to use a library in your home directory (for instance if it doesn't have a nix derivation or is a development version) then an attribute set can be passed into `agda.withPackages` with the following arguments:
+
+```
+agda.withPackages {
+  pkgs = [ agda.standard-library ];
+  homeLibraries = "path/to/my/package.agda-lib";
+}
+```
+
+Agda will not by default use these libraries. To tell agda to use the library we have some options:
+- Call `agda` with the library flag:
+```
+$ agda -l standard-library -i . MyFile.agda
```

Comment:
I tried to add a test, but it OOMs on my machine: https://github.com/turion/nixpkgs/tree/agda-rework

Reply:
```
machine # [   66.081633] [    648]     0   648 268560808    71603   757760        0             0 agda
```
Oh Agda...

Reply:
@alexarice I extended the test a bit, it still fails, but with an error now:

```
machine: output: /nix/store/_build: createDirectory: permission denied (Read-only
file system)

error: command `agda -l standard-library -i . Test2.agda` failed (exit code 1)
```
Maybe you have an idea? Feel free to add the tests if we can get them to work.

Reply:
This sounds like an error with the test machine rather than an issue with agda? Unfortunately I don't know much about testing, but I could have a look into it

Reply:
Only thing I can think of which might be happening here is that it tries to use your my-lib library, but it isn't compiled and so it tries to compile it but it can't because you've put it in the nix store

Reply:
my laptop also runs out of memory trying to compile hello world

Reply:
There is a `virtualisation.memorySize` option for tests. I might have a look tomorrow

Reply:
> Only thing I can think of which might be happening here is that it tries to use your my-lib library, but it isn't compiled and so it tries to compile it but it can't because you've put it in the nix store

Oh yes that's probably what the second error means. I'll try and change the example.

Reply:
Yeah, `agda`'s memory usage in general is abominal. Maybe only _typechecking_ instead of compiling saves some memory? (But I think that's what I'm already doing here..?)

Reply:
This does only typecheck. Did you try changing `virtualisation.memorySize`?

Reply:
I tried increasing `virtualisation.memorySize` to 2GB, but it still OOMs at ~260 MB of agda.

Reply:
The virtual machine likely takes a lot. Try bumping it to 4

Reply:
I added:
```
  virtualisation.memorySize = 8000; # Agda uses a LOT of memory
```
Still, it crashes. Weirdly, on the host I don't experience a lot of memory usage. Maybe I'm doing something wrong...

Reply:
I'll have a look in a bit

Reply:
Oh. I think I put it in the wrong place. I needed to add that to `machine`, not to the toplevel config. Now it works!

Reply:
I now have three tests, but the first (using a local library) unfortunately doesn't work yet. Can you have a look whether that's a fault of agda or the test setup? (I suspect the latter)

Reply:
@alexarice to be more precise: Adding the `homeLibraries` attribute causes agda to crash. In fact, I can reproduce this on my own machine! (I just reproduced all the steps in the `/tmp` directory.)